### PR TITLE
Add GCS token refresh for log_artifacts

### PR DESCRIPTION
**Pull request by kriscon-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/kriscon-db/mlflow/pull/13397?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13397/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13397
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Added GCP token refresh attempt to log_artifacts in the gcs_artifact_repo

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Unit test added, updated existing tests, and verified on a GCP workspace that the refresh works to get a large model registered.

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 1]
```diff
@@ -103,9 +132,10 @@ def log_artifacts(self, local_dir, artifact_path=None):
                 upload_path = posixpath.join(dest_path, rel_path)
             for f in filenames:
                 path = posixpath.join(upload_path, f)
-                gcs_bucket.blob(path, chunk_size=self._GCS_UPLOAD_CHUNK_SIZE).upload_from_filename(
-                    os.path.join(root, f), timeout=self._GCS_DEFAULT_TIMEOUT
-                )
+                # For large models, we need to speculatively retry a credential refresh
+                # and throw if it still fails.  We cannot use the built-in refresh because UC
+                # does not return a refresh token with the oauth token
+                self._retryable_log_artifact(bucket, path, os.path.join(root, f))
```
Comment (artjen) [inline mlflow/store/artifact/gcs_artifact_repo.py:138]: What's preventing us from using [_retry_with_new_creds](https://github.com/mlflow/mlflow/blob/ee4fae72121d75848d729aa3b71b2036800126a3/mlflow/store/artifact/cloud_artifact_repo.py#L88-L103) here? We can move that util function from `cloud_artifact_repo` up to `artifact_repo` so it's accessible. 
Looking at [optimized_s3_artifact_repo](https://github.com/mlflow/mlflow/blob/6814e5a6556ebf957e5f4381e3d548415f77495a/mlflow/store/artifact/optimized_s3_artifact_repo.py#L207-L227) as an example, let's make
```
def try_func(gcs_bucket):
  gcs_bucket.blob(path, chunk_size=self._GCS_UPLOAD_CHUNK_SIZE).upload_from_filename(
     file_name, timeout=self._GCS_DEFAULT_TIMEOUT
  )

def creds_func(): 
  from google.cloud.storage import Client
  from google.oauth2.credentials import Credentials

   new_token = self.credential_refresh_def()
   credentials = Credentials(new_token)
   self.client = Client(project="mlflow", credentials=credentials)
   return self._get_bucket(bucket)
```
and `og_creds = gcs_bucket`. This way we maintain the pattern with other artifact repos.

File: tests/store/artifact/test_gcs_artifact_repo.py [hunk 2]
```diff
@@ -483,3 +483,54 @@ def test_abort_multipart_upload(mock_client):
...
+        gcs_refreshed_bucket_mock.blob.return_value.upload_from_filename.side_effect = (
+            success_side_effect_func
+        )
+
+        repo = GCSArtifactRepository(
+            artifact_uri="gs://test_bucket/test_root/",
+            client=gcs_client_mock,
+            credential_refresh_def=creds_func,
+        )
+        repo._retryable_log_artifact("bucket", "path", "filename")
```
Comment (artjen) [inline tests/store/artifact/test_gcs_artifact_repo.py:525]: With the changes above, let's call `log_artifacts` instead.

File: mlflow/store/artifact/gcs_artifact_repo.py (hunk 1 above)
Comment (kriscon-db) [reply mlflow/store/artifact/gcs_artifact_repo.py:138]: So the issue im running into with that is that the other retry's only need to accept the creds, but this one will need the bucket, the path, and the filename.

Comment (artjen) [reply mlflow/store/artifact/gcs_artifact_repo.py:138]: You could pass all these arguments in as a tuple or a dict and retrieve them in the `try_func`.

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 3]
```diff
@@ -88,13 +106,39 @@ def log_artifact(self, local_file, artifact_path=None):
         blob = gcs_bucket.blob(dest_path, chunk_size=self._GCS_UPLOAD_CHUNK_SIZE)
         blob.upload_from_filename(local_file, timeout=self._GCS_DEFAULT_TIMEOUT)
 
+    def _retryable_log_artifact(self, bucket, path, file_name):
```
Comment (artjen) [inline mlflow/store/artifact/gcs_artifact_repo.py:109]: Do we still need this function if we're using `_retry_with_new_creds`?

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 4]
```diff
@@ -19,11 +20,16 @@
...
+from mlflow.store.artifact.artifact_repo import (
+    ArtifactRepository,
+    MultipartUploadMixin,
+    _retry_with_new_creds
+)
 from mlflow.utils.file_utils import relative_path_to_artifact_path
 
 GCSMPUArguments = namedtuple("GCSMPUArguments", ["transport", "url", "headers", "content_type"])
 
+_logger = logging.getLogger(__name__)
```
Comment (artjen) [inline mlflow/store/artifact/gcs_artifact_repo.py:32]: If we remove `_retryable_log_artifact`, please remove this as well.

File: tests/store/artifact/test_gcs_artifact_repo.py [hunk 5]
```diff
@@ -483,3 +483,61 @@ def test_abort_multipart_upload(mock_client):
...
+
+        data = tmp_path.joinpath("data")
+        data.mkdir()
+        subd = data.joinpath("subdir")
+        subd.mkdir()
+        subd.joinpath("a.txt").write_text("A")
+
+        try:
+            repo.log_artifacts(subd)
+        except Exception:
```
Comment (harupy) [inline tests/store/artifact/test_gcs_artifact_repo.py:531]: We shouldn't use try-catch in a test. Can we use `pytest.raises` instead if `log_artifacts` should fail?

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 6]
```diff
@@ -88,13 +106,39 @@ def log_artifact(self, local_file, artifact_path=None):
...
+        except Exception as e:
+            from google.cloud.storage import Client
+            from google.oauth2.credentials import Credentials
+
+            _logger.info(
+                "Failed to complete upload request, possibly due to credential expiration."
+                f" Attempting to refresh credentials and trying again... (Error: {e})"
+            )
+            if self.credential_refresh_def is not None:
+                new_token = self.credential_refresh_def()
```
Comment (harupy) [inline mlflow/store/artifact/gcs_artifact_repo.py:124]:
```suggestion

            if self.credential_refresh_def is not None:
                _logger.info(
                "Failed to complete upload request, possibly due to credential expiration."
                f" Attempting to refresh credentials and trying again... (Error: {e})"
            )
                new_token = self.credential_refresh_def()
```

File: mlflow/store/artifact/artifact_repo.py [hunk 7]
```diff
@@ -38,6 +38,22 @@ def _truncate_error(err: str, max_length: int = 10_000) -> str:
     half = max_length // 2
     return err[:half] + "\n\n*** Error message is too long, truncated ***\n\n" + err[-half:]
 
+def _retry_with_new_creds(try_func, creds_func, og_creds=None):
```
Comment (harupy) [inline mlflow/store/artifact/artifact_repo.py:41]:
```suggestion
def _retry_with_new_creds(try_func, creds_func, orig_creds=None):
```

File: mlflow/store/artifact/artifact_repo.py [hunk 8]
```diff
@@ -38,6 +38,22 @@ def _truncate_error(err: str, max_length: int = 10_000) -> str:
...
+    credentials using creds_func. If the try_func throws, then try again with new credentials
+    provided by creds_func.
+    """
+    try:
+        first_creds = creds_func() if og_creds is None else og_creds
+        return try_func(first_creds)
+    except Exception as e:
+        _logger.info(
+            "Failed to complete request, possibly due to credential expiration."
+            f" Refreshing credentials and trying again... (Error: {e})"
```
Comment (harupy) [inline mlflow/store/artifact/artifact_repo.py:53]:
```suggestion
            "Failed to complete request, possibly due to credential expiration (Error: {e})."
            f" Refreshing credentials and trying again..."
```

File: mlflow/store/artifact/gcs_artifact_repo.py (hunk 6 above)
Comment (kriscon-db) [reply mlflow/store/artifact/gcs_artifact_repo.py:124]: This logger message is to indicate that we failed due to a potential credential expiration and that we are attempting to refresh again.  I don't think it should be in the if block.

**Review by harupy — APPROVED** [review]

File: tests/store/artifact/test_gcs_artifact_repo.py [hunk 9]
```diff
@@ -483,3 +483,61 @@ def test_abort_multipart_upload(mock_client):
...
+        if throw:
+            assert gcs_bucket_mock.blob.call_count == 1
+            assert gcs_refreshed_bucket_mock.blob.call_count == 1
+            assert mock_gcs_client_factory.call_count == 1
+            assert mock_gcs_credentials_factory.call_count == 1
+        else:
+            assert gcs_bucket_mock.blob.call_count == 1
+            assert gcs_refreshed_bucket_mock.blob.call_count == 0
+            assert mock_gcs_client_factory.call_count == 0
+            assert mock_gcs_credentials_factory.call_count == 0
```
Comment (harupy) [inline tests/store/artifact/test_gcs_artifact_repo.py:543]: /ai rewrite this using `assert_called_once` and `assert_not_called`

File: mlflow/store/artifact/artifact_repo.py [hunk 10]
```diff
@@ -39,6 +39,24 @@ def _truncate_error(err: str, max_length: int = 10_000) -> str:
...
 
 
+def _retry_with_new_creds(try_func, creds_func, orig_creds=None):
+    """
+    Attempt the try_func with the original credentials (og_creds) if provided, or by generating the
+    credentials using creds_func. If the try_func throws, then try again with new credentials
+    provided by creds_func.
+    """
+    try:
+        first_creds = creds_func() if orig_creds is None else orig_creds
```
Comment (harupy) [inline mlflow/store/artifact/artifact_repo.py:49]: is there any chance that `creds_func()` throws?

File: mlflow/utils/_unity_catalog_utils.py [hunk 11]
```diff
@@ -221,8 +221,20 @@ def azure_credential_refresh():
         from mlflow.store.artifact.gcs_artifact_repo import GCSArtifactRepository
 
         credentials = Credentials(scoped_token.gcp_oauth_token.oauth_token)
+
+        def gcp_credential_refresh():
+            new_scoped_token = base_credential_refresh_def()
+            new_gcp_creds = new_scoped_token.gcp_oauth_token
+            return {
+                "oauth_token": new_gcp_creds.oauth_token,
+            }
```
Comment (harupy) [inline mlflow/utils/_unity_catalog_utils.py:230]: do we need to return a dict? can we return `new_gcp_creds.oauth_token`?

**Review by artjen — APPROVED** [review]

File: mlflow/store/artifact/artifact_repo.py (hunk 10 above)
Comment (kriscon-db) [reply mlflow/store/artifact/artifact_repo.py:49]: Yes but that would be a UC error and we should just fail the upload at that point.

**Pull request merged**